promotion/datastore.go

```diff
@@ -1696,6 +1849,44 @@ func (pg *Postgres) RunNextDrainRetryJob(ctx context.Context, worker DrainRetryW
 	}
 }
 
+// RunNextFetchRewardGrantsJob - runs fetch reward grant jobs
+func (pg *Postgres) RunNextFetchRewardGrantsJob(ctx context.Context, worker SwapRewardsWorker) error {
+
+	service, _ := InitService(ctx, pg, nil, nil)
+
+	grant, msg, err := worker.FetchRewardsGrants(ctx)
+	if err != nil {
+		return fmt.Errorf("fetch rewards grants job: failed to retrieve grant information: %w", err)
+	}
+
+	cohort := "control"
+	//maybe this call should be changed to just rely on the right status codes and only do 1 post request
+	_, err = service.GetOrCreateIssuer(ctx, grant.PromotionID, cohort)
+	if err != nil {
+		return err
+	}
+
+	claimType := "swap"
+	statement := `
+			insert into claims (promotion_id, wallet_id, address_id, transaction_key, approximate_value, legacy_claimed, claim_type)
+			values ($1, $2, $3, $4, $5, false, $6)
+			`
+
+	_, err = pg.RawDB().Exec(statement, grant.PromotionID, swapSentinelWalletID, grant.AddressID, grant.TransactionKey, grant.RewardAmount, claimType)
+
+	if err != nil {
+		err = fmt.Errorf("fetch rewards grants job: failed to update claim WalletID: %s PromotionID: %s with err: %w", grant.AddressID, grant.PromotionID, err)
+		return err
+	}
+
+	if err := service.kafkaGrantRewardsReader.CommitMessages(ctx, msg); err != nil {
+		err = fmt.Errorf("failed to commit kafka offset AddressID: %s PromotionID: %s with err: %w", grant.AddressID, grant.PromotionID, err)
+		return err
+	}
+
+	return nil
+}
```

Comment:
Does this logic handle the case when there are two messages with the same `transactions_key`? I think Postgres would bubble up an error and we'd have to handle that, then commit the message to continue consuming

Reply:
It does now https://github.com/brave-intl/bat-go/pull/1213/commits/36af29ab409da8c030a17b2330181d5423979a6d good call-out